src/server.rs

```diff
@@ -475,11 +475,26 @@ pub trait Output {
     }
 
     fn notify(&self, message: &str) {
-        let output = serde_json::to_string(
-            &NotificationMessage::new(message.to_owned(), ())
-        ).unwrap();
+        let output = serde_json::to_string(&NotificationMessage::new(message.to_owned(), ()))
+            .unwrap();
         self.response(output);
     }
+
+    fn notify_begin(&self) {
+        self.notify("rustDocument/diagnosticsBegin");
+    }
+
+    fn notify_end(&self) {
+        self.notify("rustDocument/diagnosticsEnd");
+    }
+
+    fn notify_biuld_results(&self,
+                            notifications: &Vec<NotificationMessage<PublishRustDiagnosticsParams>>) {
+        for notification in notifications {
```

Comment:
`Output` is a fairly abstract trait and I don't think it should know about the specifics of notifications like this. If you look at the `success` and `failure` methods, you can see we just take a message string and id, we let the client do this kind of processing.